astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+                               int const nanosecond)
+    : hour_(hour),
+      minute_(minute),
+      second_(second),
+      nanosecond_(nanosecond) {}
+
+constexpr TimeOfDay const& TimeOfDay::checked() const {
+  return CHECKING(
+      (hour_ == 24 && minute_ == 0 && second_ == 0 && nanosecond_ == 0) ||
+          ((nanosecond_ >= 0 && nanosecond_ <= 999'999'999) &&
+           ((hour_ == 23 && minute_ == 59 && second_ == 60) ||
+            (hour_ >= 0 && hour_ <= 23 && minute_ >= 0 && minute_ <= 59 &&
+             second_ >= 0 && second_ <= 59))),
+      *this);
+}
+
+// Implementation of class DateTime.
+
+constexpr Date const& DateTime::date() const {
+  return date_;
+}
+
+constexpr TimeOfDay const& DateTime::time() const {
+  return time_;
+}
+
+constexpr DateTime DateTime::normalized_end_of_day() const {
+  return time_.is_end_of_day()
+             ? DateTime(date_.next_day(), TimeOfDay::hhmmss_ns(00'00'00, 0))
+             : *this;
+}
+
+constexpr DateTime::DateTime(Date const date, TimeOfDay const time)
+    : date_(date),
+      time_(time) {}
+
+constexpr DateTime const& DateTime::checked() const {
+  return CHECKING(
+      !time_.is_leap_second() ||
+          (date_.day() == month_length(date_.year(), date_.month()) &&
+           (date_.month() == 6 || date_.month() == 12 || date_.month() == 3 ||
+            date_.month() == 9)),
+      *this);
+}
+
+// Date parsing.
+
+struct DateStringInfo {
+  constexpr DateStringInfo Fill() const {
+    return read == size
+               ? *this
+               : string[read] == '-'
+                     ? CHECKING(
+                           hyphens < 2,
+                           hyphens == 0
+                               ? DateStringInfo({string,
+                                                 size,
+                                                 read + 1,
+                                                 digits,
+                                                 digit_count,
+                                                 hyphens + 1,
+                                                 /*first_hyphen_index=*/read,
+                                                 second_hyphen_index,
+                                                 has_w,
+                                                 w_index}).Fill()
+                               : DateStringInfo({string,
+                                                 size,
+                                                 read + 1,
+                                                 digits,
+                                                 digit_count,
+                                                 hyphens + 1,
+                                                 first_hyphen_index,
+                                                 /*second_hyphen_index=*/read,
+                                                 has_w,
+                                                 w_index})).Fill()
+                     : string[read] == 'W'
```

Comment:
Commentscommentcomments.  Explain what format is being parsed by each alternative.